// @tsed/di/lib/class/Provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@tsed/core");
const interfaces_1 = require("../interfaces");
const ProviderType_1 = require("../interfaces/ProviderType");
class Provider {
    constructor(token) {
        /**
         *
         */
        this.type = ProviderType_1.ProviderType.PROVIDER;
        /**
         *
         */
        this.injectable = true;
        this.provide = token;
        this.useClass = token;
    }
    /**
     *
     * @returns {any}
     */
    get provide() {
        return this._provide;
    }
    /**
     *
     * @param value
     */
    set provide(value) {
        this._provide = core_1.isClass(value) ? core_1.getClass(value) : value;
    }
    /**
     *
     * @returns {Type<T>}
     */
    get useClass() {
        return this._useClass;
    }
    /**
     * Create a new store if the given value is a class. Otherwise the value is ignored.
     * @param value
     */
    set useClass(value) {
        if (core_1.isClass(value)) {
            this._useClass = core_1.getClass(value);
            this._store = core_1.Store.from(value);
        }
    }
    /**
     *
     * @returns {string}
     */
    get className() {
        return this.name;
    }
    /**
     *
     */
    get name() {
        return core_1.nameOf(this.provide);
    }
    /**
     *
     * @returns {Store}
     */
    get store() {
        return this._store;
    }
    /**
     * Get the scope of the provider.
     * @returns {boolean}
     */
    get scope() {
        return this._store ? this.store.get("scope") : this._scope;
    }
    /**
     * Change the scope value of the provider.
     * @param scope
     */
    set scope(scope) {
        this._store ? this.store.set("scope", scope) : this._scope;
    }
    /**
     *
     */
    clone() {
        const provider = new (core_1.getClass(this))(this.provide);
        core_1.getKeys(this).forEach(key => {
            provider[key] = this[key];
        });
        return provider;
    }
    toString() {
        return `Token:${this.name}`;
    }
}
tslib_1.__decorate([
    core_1.Enumerable(),
    tslib_1.__metadata("design:type", Object)
], Provider.prototype, "type", void 0);
tslib_1.__decorate([
    core_1.Enumerable(),
    tslib_1.__metadata("design:type", Boolean)
], Provider.prototype, "injectable", void 0);
tslib_1.__decorate([
    core_1.Enumerable(),
    tslib_1.__metadata("design:type", Object)
], Provider.prototype, "instance", void 0);
tslib_1.__decorate([
    core_1.Enumerable(),
    tslib_1.__metadata("design:type", Array)
], Provider.prototype, "deps", void 0);
tslib_1.__decorate([
    core_1.Enumerable(),
    tslib_1.__metadata("design:type", Function)
], Provider.prototype, "useFactory", void 0);
tslib_1.__decorate([
    core_1.Enumerable(),
    tslib_1.__metadata("design:type", Object)
], Provider.prototype, "useValue", void 0);
tslib_1.__decorate([
    core_1.NotEnumerable(),
    tslib_1.__metadata("design:type", Object)
], Provider.prototype, "_provide", void 0);
tslib_1.__decorate([
    core_1.NotEnumerable(),
    tslib_1.__metadata("design:type", core_1.Type)
], Provider.prototype, "_useClass", void 0);
tslib_1.__decorate([
    core_1.NotEnumerable(),
    tslib_1.__metadata("design:type", Object)
], Provider.prototype, "_instance", void 0);
tslib_1.__decorate([
    core_1.NotEnumerable(),
    tslib_1.__metadata("design:type", String)
], Provider.prototype, "_scope", void 0);
tslib_1.__decorate([
    core_1.NotEnumerable(),
    tslib_1.__metadata("design:type", core_1.Store)
], Provider.prototype, "_store", void 0);
tslib_1.__decorate([
    core_1.Enumerable(),
    tslib_1.__metadata("design:type", core_1.Type),
    tslib_1.__metadata("design:paramtypes", [core_1.Type])
], Provider.prototype, "useClass", null);
tslib_1.__decorate([
    core_1.Enumerable(),
    tslib_1.__metadata("design:type", String),
    tslib_1.__metadata("design:paramtypes", [String])
], Provider.prototype, "scope", null);
exports.Provider = Provider;
