// @tsed/common/lib/server/utils/createExpressApplication.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@tsed/common");
const Express = require("express");
const mvc_1 = require("../../mvc");
const expressApplication_1 = require("../decorators/expressApplication");
function createExpressApplication(injector) {
    /* await */
    injector.forkProvider(expressApplication_1.ExpressApplication);
}
exports.createExpressApplication = createExpressApplication;
common_1.registerProvider({
    provide: expressApplication_1.ExpressApplication,
    deps: [common_1.InjectorService],
    scope: common_1.ProviderScope.SINGLETON,
    global: true,
    useFactory(injector) {
        const expressApp = Express();
        const originalUse = expressApp.use;
        expressApp.use = function (...args) {
            args = args.map(arg => {
                if (injector.has(arg)) {
                    arg = mvc_1.HandlerBuilder.from(arg).build(injector);
                }
                return arg;
            });
            return originalUse.call(this, ...args);
        };
        return expressApp;
    }
});
