// @tsed/common/lib/config/services/ServerSettingsService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
var ServerSettingsService_1;
const core_1 = require("@tsed/core");
const di_1 = require("@tsed/di");
const ts_log_debug_1 = require("ts-log-debug");
const index_1 = require("../constants/index");
const rootDir = process.cwd();
/**
 * `ServerSettingsService` contains all information about [ServerLoader](/api/common/server/components/ServerLoader.md) configuration.
 */
let ServerSettingsService = ServerSettingsService_1 = class ServerSettingsService {
    constructor() {
        this.map = new Map();
        this.rootDir = rootDir;
        this.env = process.env.NODE_ENV || core_1.Env.DEV;
        this.port = 8080;
        this.httpsPort = 8000;
        this.version = "1.0.0";
        this.uploadDir = "${rootDir}/uploads";
        this.controllerScope = di_1.ProviderScope.SINGLETON;
        this.logger = {
            debug: false,
            level: "info",
            logRequest: true,
            jsonIndentation: this.env === core_1.Env.PROD ? 0 : 2
        };
        this.errors = {
            headerName: "errors"
        };
        this.mount = {
            "/rest": "${rootDir}/controllers/**/*.ts"
        };
        this.exclude = ["**/*.spec.ts", "**/*.spec.js"];
        this.componentsScan = ["${rootDir}/mvc/**/*.ts", "${rootDir}/services/**/*.ts", "${rootDir}/converters/**/*.ts"];
    }
    /**
     *
     * @returns {any}
     */
    get version() {
        return this.map.get("version");
    }
    set version(v) {
        this.map.set("version", v);
    }
    /**
     *
     * @returns {any}
     */
    get rootDir() {
        return this.map.get("rootDir");
    }
    /**
     *
     * @param value
     */
    set rootDir(value) {
        this.map.set("rootDir", value);
    }
    /**
     *
     * @param value
     */
    set port(value) {
        this.httpPort = value;
    }
    /**
     *
     * @param value
     */
    get httpsOptions() {
        return this.map.get("httpsOptions");
    }
    /**
     *
     * @param value
     */
    set httpsOptions(value) {
        this.map.set("httpsOptions", value);
    }
    /**
     *
     * @returns {undefined|any}
     */
    get httpPort() {
        return this.map.get("httpPort");
    }
    /**
     *
     * @param value
     */
    set httpPort(value) {
        this.map.set("httpPort", value);
    }
    /**
     *
     * @returns {undefined|any}
     */
    get httpsPort() {
        return this.map.get("httpsPort");
    }
    /**
     *
     * @param value
     */
    set httpsPort(value) {
        this.map.set("httpsPort", value);
    }
    /**
     *
     * @returns {string}
     */
    get uploadDir() {
        return this.map.get("uploadDir");
    }
    /**
     *
     * @param value
     */
    set uploadDir(value) {
        this.map.set("uploadDir", value);
    }
    /**
     *
     * @returns {Map<string, any>}
     */
    get env() {
        return this.map.get("env");
    }
    /**
     *
     * @param value
     */
    set env(value) {
        this.map.set("env", value);
    }
    /**
     *
     * @returns {undefined|any}
     */
    get mount() {
        return this.map.get("mount") || {};
    }
    /**
     *
     * @param value
     */
    set mount(value) {
        this.map.set("mount", value);
    }
    /**
     *
     * @returns {undefined|any}
     */
    get componentsScan() {
        return this.map.get("componentsScan") || [];
    }
    /**
     *
     * @param value
     */
    set componentsScan(value) {
        this.map.set("componentsScan", value);
    }
    /**
     *
     * @returns {undefined|any}
     */
    get statics() {
        return this.map.get("statics") || this.map.get("serveStatic") || {};
    }
    /**
     *
     * @param value
     */
    set statics(value) {
        this.map.set("statics", value);
    }
    /**
     *
     * @param value
     */
    /* istanbul ignore next */
    get serveStatics() {
        return this.statics;
    }
    /* istanbul ignore next */
    set serveStatics(value) {
        this.statics = value;
    }
    /**
     *
     * @returns {undefined|any}
     */
    get acceptMimes() {
        return this.map.get("acceptMimes") || ["application/json"];
    }
    /**
     *
     * @param value
     */
    set acceptMimes(value) {
        this.map.set("acceptMimes", value || []);
    }
    /**
     *
     * @returns {boolean}
     */
    get debug() {
        return this.logger.level === "info";
    }
    /**
     *
     * @param {boolean} debug
     */
    set debug(debug) {
        this.logger = Object.assign({}, this.logger, { level: debug ? "debug" : "info" });
    }
    /**
     *
     * @returns {IRouterSettings}
     */
    get routers() {
        return this.map.get("routers") || {};
    }
    /**
     *
     * @param {IRouterSettings} options
     */
    set routers(options) {
        this.map.set("routers", options);
    }
    /**
     *
     * @returns {boolean}
     */
    get validationModelStrict() {
        const value = this.map.get("validationModelStrict");
        return value === undefined ? true : value;
    }
    /**
     *
     * @param {boolean} value
     */
    set validationModelStrict(value) {
        this.map.set("validationModelStrict", value);
    }
    get logger() {
        return this.map.get("logger") || {};
    }
    set logger(value) {
        const requestFields = this.get("logRequestFields");
        const logger = Object.assign({ requestFields }, this.logger, value);
        logger.debug = logger.level === "debug";
        this.map.set("logger", logger);
        this.map.set("debug", logger.debug);
        if (logger.format) {
            ts_log_debug_1.$log.appenders.set("stdout", {
                type: "stdout",
                levels: ["info", "debug"],
                layout: {
                    type: "pattern",
                    pattern: logger.format
                }
            });
            ts_log_debug_1.$log.appenders.set("stderr", {
                levels: ["trace", "fatal", "error", "warn"],
                type: "stderr",
                layout: {
                    type: "pattern",
                    pattern: logger.format
                }
            });
        }
    }
    get exclude() {
        return this.map.get("exclude") || [];
    }
    set exclude(exclude) {
        this.map.set("exclude", exclude);
    }
    get controllerScope() {
        return this.map.get("scope");
    }
    set controllerScope(scope) {
        this.map.set("scope", scope);
    }
    /**
     *
     * @returns {IRouterSettings}
     */
    get errors() {
        return this.map.get("errors") || {};
    }
    /**
     *
     * @param {IRouterSettings} options
     */
    set errors(options) {
        this.map.set("errors", options);
    }
    /**
     *
     * @param target
     * @returns {any}
     */
    static getMetadata(target) {
        return core_1.Metadata.getOwn(index_1.SERVER_SETTINGS, target);
    }
    /**
     *
     * @param addressPort
     * @returns {{address: string, port: number}}
     */
    static buildAddressAndPort(addressPort) {
        let address = "0.0.0.0";
        let port = addressPort;
        if (typeof addressPort === "string" && addressPort.indexOf(":") > -1) {
            [address, port] = addressPort.split(":");
            port = +port;
        }
        return { address, port: port };
    }
    /**
     *
     * @param callbackfn
     * @param thisArg
     */
    forEach(callbackfn, thisArg) {
        return this.map.forEach(callbackfn, thisArg);
    }
    /**
     *
     * @param propertyKey
     * @param value
     */
    set(propertyKey, value) {
        if (typeof propertyKey === "string") {
            core_1.setValue(propertyKey, value, this.map);
        }
        else {
            const self = this;
            Object.keys(propertyKey).forEach(key => {
                const descriptor = Object.getOwnPropertyDescriptor(ServerSettingsService_1.prototype, key);
                if (descriptor && ["set", "map"].indexOf(key) === -1) {
                    self[key] = propertyKey[key];
                }
                else {
                    this.set(key, propertyKey[key]);
                }
            });
            this.forEach((value, key) => {
                this.map.set(key, this.resolve(value));
            });
        }
        return this;
    }
    /**
     *
     * @param propertyKey
     * @returns {undefined|any}
     */
    get(propertyKey) {
        return this.resolve(core_1.getValue(propertyKey, this.map));
    }
    /**
     *
     * @param value
     * @returns {any}
     */
    resolve(value) {
        if (typeof value === "object" && value !== null) {
            Object.keys(value).forEach((k, i, m) => {
                value[k] = this.resolve(value[k]);
            });
            return value;
        }
        if (typeof value === "string") {
            return value.replace(/\${rootDir}/, this.rootDir);
        }
        return value;
    }
    /**
     *
     * @returns {string|number}
     */
    getHttpPort() {
        return ServerSettingsService_1.buildAddressAndPort(this.map.get("httpPort"));
    }
    /**
     *
     * @param settings
     */
    setHttpPort(settings) {
        this.map.set("httpPort", `${settings.address}:${settings.port}`);
    }
    /**
     *
     * @returns {string|number}
     */
    getHttpsPort() {
        return ServerSettingsService_1.buildAddressAndPort(this.map.get("httpsPort"));
    }
    /**
     *
     * @param settings
     */
    setHttpsPort(settings) {
        this.map.set("httpsPort", `${settings.address}:${settings.port}`);
    }
};
ServerSettingsService = ServerSettingsService_1 = tslib_1.__decorate([
    di_1.Injectable({
        scope: di_1.ProviderScope.SINGLETON,
        global: true
    }),
    tslib_1.__metadata("design:paramtypes", [])
], ServerSettingsService);
exports.ServerSettingsService = ServerSettingsService;
