// ts-httpexceptions/src/core/Exception.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Exception extends Error {
    /**
     *
     * @param status
     * @param message
     * @param innerException
     */
    constructor(status, message, innerException) {
        super(message);
        /**
         * Exception base name
         * @type {string}
         */
        this.name = "HTTP_EXCEPTION";
        /**
         * Exception type
         * @type {string}
         */
        this.type = "HTTP_EXCEPTION";
        this.status = status;
        this.message = message || "";
        if (innerException) {
            if (innerException instanceof Error) {
                this.innerException = innerException;
                this.message = this.message + ", innerException: " + this.innerException.message;
            }
            else if (typeof innerException === "string") {
                this.innerException = new Error(innerException);
                this.message = this.message + ", innerException: " + this.innerException.message;
            }
            else {
                this.innerException = innerException;
                this.message = this.message + ", innerException: " + this.innerException;
            }
        }
        this.message = (this.message + " ").trim();
    }
    toString() {
        // console.warn(this.type + ": " + this.status + " " + this.name  + " => " + this.message)
        return (this.name + "(" + this.status + "): " + this.message + " ").trim();
    }
}
exports.Exception = Exception;
class HTTPException extends Exception {
}
exports.HTTPException = HTTPException;
