// @tsed/common/lib/converters/errors/RequiredPropertyError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@tsed/core");
const ts_httpexceptions_1 = require("ts-httpexceptions");
/**
 * @private
 */
class RequiredPropertyError extends ts_httpexceptions_1.BadRequest {
    constructor(target, propertyName) {
        super(RequiredPropertyError.buildMessage(target, propertyName));
        this.errors = [
            {
                dataPath: "",
                keyword: "required",
                message: `should have required property '${String(propertyName)}'`,
                modelName: core_1.nameOf(target),
                params: {
                    missingProperty: propertyName
                },
                schemaPath: "#/required"
            }
        ];
    }
    /**
     *
     * @returns {string}
     * @param target
     * @param propertyName
     */
    static buildMessage(target, propertyName) {
        return `Property ${propertyName} on class ${core_1.nameOf(target)} is required.`;
    }
}
exports.RequiredPropertyError = RequiredPropertyError;
