// @tsed/common/lib/filters/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const FilterRegistry_1 = require("../registries/FilterRegistry");
exports.PARAM_METADATA = "tsed:params:metadata";
/**
 * Metadata key
 * @private
 * @type {string}
 */
exports.EXPRESS_NEXT_FN = Symbol("next");
/**
 * Metadata key
 * @private
 * @type {string}
 */
exports.EXPRESS_ERR = Symbol("err");
/**
 * Metadata key
 * @private
 * @type {string}
 */
exports.EXPRESS_REQUEST = Symbol("request");
/**
 * Metadata key
 * @private
 * @type {string}
 */
exports.EXPRESS_RESPONSE = Symbol("response");
/**
 * Metadata key
 * @private
 * @type {string}
 */
exports.RESPONSE_DATA = Symbol("responseData");
/**
 * Metadata key
 * @private
 * @type {string}
 */
exports.ENDPOINT_INFO = Symbol("endpointInfo");
/**
 * Response PreHandler
 */
FilterRegistry_1.FilterPreHandlers.set(exports.EXPRESS_RESPONSE, locals => locals.response);
/**
 * Request PreHandler
 */
FilterRegistry_1.FilterPreHandlers.set(exports.EXPRESS_REQUEST, locals => locals.request);
/**
 * Next PreHandler
 */
FilterRegistry_1.FilterPreHandlers.set(exports.EXPRESS_NEXT_FN, locals => locals.next);
/**
 * Express error PreHandler
 */
FilterRegistry_1.FilterPreHandlers.set(exports.EXPRESS_ERR, locals => locals.err);
/**
 * EndpointInfo PreHandler
 */
FilterRegistry_1.FilterPreHandlers.set(exports.ENDPOINT_INFO, locals => {
    const op = FilterRegistry_1.FilterPreHandlers.get(exports.EXPRESS_REQUEST);
    return op(locals).ctx.endpoint;
});
/**
 * ResponseData PreHandler
 */
FilterRegistry_1.FilterPreHandlers.set(exports.RESPONSE_DATA, locals => {
    const op = FilterRegistry_1.FilterPreHandlers.get(exports.EXPRESS_REQUEST);
    return op(locals).ctx.data;
});
