// @tsed/common/lib/mvc/class/HandlerMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@tsed/core");
const constants_1 = require("../../filters/constants");
const ParamRegistry_1 = require("../../filters/registries/ParamRegistry");
const HandlerType_1 = require("../interfaces/HandlerType");
class HandlerMetadata {
    constructor(options) {
        this.injectable = false;
        this.type = HandlerType_1.HandlerType.FUNCTION;
        this.hasErrorParam = false;
        this.hasNextFunction = false;
        const { target, token, method, type = HandlerType_1.HandlerType.FUNCTION } = options;
        this.type = type;
        this.handler = method ? target.prototype[method] : target;
        if (method) {
            this.target = target;
            this.token = token;
            this.methodClassName = method;
            this.method = method;
            this.hasNextFunction = this.hasParamType(constants_1.EXPRESS_NEXT_FN);
            this.hasErrorParam = this.hasParamType(constants_1.EXPRESS_ERR);
            this.injectable = (core_1.Metadata.get(constants_1.PARAM_METADATA, target, method) || []).length > 0;
        }
        if (!this.injectable) {
            this.hasErrorParam = this.handler.length === 4;
            this.hasNextFunction = this.handler.length >= 3;
        }
    }
    get services() {
        if (this.injectable) {
            return this.getParams();
        }
        const parameters = [{ service: constants_1.EXPRESS_REQUEST }, { service: constants_1.EXPRESS_RESPONSE }];
        if (this.hasErrorParam) {
            parameters.unshift({ service: constants_1.EXPRESS_ERR });
        }
        if (this.hasNextFunction) {
            parameters.push({ service: constants_1.EXPRESS_NEXT_FN });
        }
        return parameters;
    }
    getParams() {
        return ParamRegistry_1.ParamRegistry.getParams(this.target, this.methodClassName) || [];
    }
    hasParamType(paramType) {
        return this.getParams().findIndex(p => p.service === paramType) > -1;
    }
}
exports.HandlerMetadata = HandlerMetadata;
