// @tsed/di/lib/errors/InjectionError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@tsed/core");
class InjectionError extends Error {
    constructor(token, origin) {
        super(core_1.isString(origin) ? origin : "");
        this.name = "INJECTION_ERROR";
        this.tokens = [];
        this.tokens = [token];
        if (origin) {
            if (core_1.isString(origin)) {
                this.origin = {
                    message: origin,
                    stack: this.stack
                };
            }
            else {
                if (origin.tokens) {
                    this.tokens = this.tokens.concat(origin.tokens);
                    this.origin = origin.origin;
                }
                else {
                    this.origin = origin;
                }
            }
        }
        const originMessage = this.origin ? "\nOrigin: " + this.origin.message : "";
        const tokensMessage = this.tokens.map(token => core_1.nameOf(token)).join(" > ");
        this.message = `Injection failed on ${tokensMessage}${originMessage}`;
    }
}
exports.InjectionError = InjectionError;
