// @tsed/common/lib/filters/decorators/bodyParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BodyParamsFilter_1 = require("../components/BodyParamsFilter");
const ParamTypes_1 = require("../interfaces/ParamTypes");
const ParamRegistry_1 = require("../registries/ParamRegistry");
/**
 * BodyParams return the value from [request.body](http://expressjs.com/en/4x/api.html#req.body) object.
 *
 * #### Example
 *
 * ```typescript
 * @Controller('/')
 * class MyCtrl {
 *    @Post('/')
 *    create(@BodyParams() body: any) {
 *       console.log('Entire body', body);
 *    }
 *
 *    @Post('/')
 *    create(@BodyParams('id') id: string) {
 *       console.log('ID', id);
 *    }
 *
 *    @Post('/')
 *    create(@BodyParams('user') user: User) { // with deserialization
 *       console.log('user', user);
 *    }
 *
 *    @Post('/')
 *    create(@BodyParams('users', User) users: User[]) { // with deserialization
 *       console.log('users', users);
 *    }
 * }
 * ```
 * > For more information on deserialization see [converters](/docs/converters.md) page.
 *
 * @param expression The path of the property to get.
 * @param useType The type of the class that to be used to deserialize the data.
 * @decorator
 * @returns {Function}
 */
function BodyParams(expression, useType) {
    return ParamRegistry_1.ParamRegistry.decorate(BodyParamsFilter_1.BodyParamsFilter, {
        expression,
        useType,
        useConverter: true,
        useValidation: true,
        paramType: ParamTypes_1.ParamTypes.BODY
    });
}
exports.BodyParams = BodyParams;
