// @tsed/common/lib/filters/decorators/queryParams.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const QueryParamsFilter_1 = require("../components/QueryParamsFilter");
const ParamTypes_1 = require("../interfaces/ParamTypes");
const ParamRegistry_1 = require("../registries/ParamRegistry");
/**
 * QueryParams return the value from [request.query](http://expressjs.com/en/4x/api.html#req.query) object.
 *
 * #### Example
 *
 * ```typescript
 * @Controller('/')
 * class MyCtrl {
 *    @Get('/')
 *    get(@QueryParams() query: any) {
 *       console.log('Entire query', query);
 *    }
 *
 *    @Get('/')
 *    get(@QueryParams('id') id: string) {
 *       console.log('ID', id);
 *    }
 *
 *    @Get('/')
 *    get(@QueryParams('user') user: User) { // with deserialization
 *       console.log('user', user);
 *    }
 *
 *    @Get('/')
 *    get(@QueryParams('users', User) users: User[]) { // with deserialization
 *       console.log('users', users);
 *    }
 * }
 * ```
 * > For more information on deserialization see [converters](/docs/converters.md) page.
 *
 * @param expression The path of the property to get.
 * @param useType The type of the class that to be used to deserialize the data.
 * @decorator
 * @returns {Function}
 */
function QueryParams(expression, useType) {
    return ParamRegistry_1.ParamRegistry.decorate(QueryParamsFilter_1.QueryParamsFilter, {
        expression,
        useType,
        useConverter: true,
        useValidation: true,
        paramType: ParamTypes_1.ParamTypes.QUERY
    });
}
exports.QueryParams = QueryParams;
